harness/determined/ipc.py

```diff
@@ -439,3 +443,216 @@ def barrier(self, message: Any = None) -> Any:
+            self.listener.close()
+            self.listener = None
+        if self.sel:
+            self.sel.close()
+            self.sel = None
+
+    def __enter__(self) -> "PIDServer":
+        return self.start()
+
+    def __exit__(self, *_: Any) -> None:
+        self.close()
+
+    def get_port(self) -> int:
+        assert self.port is not None, "must start first"
+        return self.port
+
+    def handle_listener(self, mask: int) -> None:
+        """
+        Handle an event on a listener socket (aka, accept a connection).
+        """
+        assert self.sel
+        assert self.listener
+        if mask & selectors.EVENT_READ:
+            conn, _ = self.listener.accept()
+            # We never write anything.
+            conn.shutdown(socket.SHUT_WR)
+            # First, receive the initial PID for this conn.  Should be nearly instant.
+            buf = b""
+            while b"\n" not in buf:
+                data = conn.recv(4096)
+                if not data:
+                    raise ValueError("pid_client did not deliver a PID!")
+                buf += data
+            pid_buf, data_buf = buf.split(b"\n", 1)
+            pid = int(pid_buf)
+            self.pids.append(pid)
+            self.conns[conn] = pid
+            # Now listen for this connection to gracefully shut down (eventually)
+            conn.setblocking(False)
+            self.sel.register(conn, selectors.EVENT_READ)
+            if len(self.pids) == self.num_clients:
+                # That the last connection, close the listener.
+                self.sel.unregister(self.listener)
+                self.listener.close()
+                self.listener = None
+            if data_buf:
+                # We received a message in the same packet as the PID, simulate an EVENT_READ.
+                self.handle_conn(conn, mask=0, data=data_buf)
+        else:
+            raise ValueError("listener failed")
+
+    def handle_conn(self, conn: socket.socket, mask: int, data: Optional[bytes] = None) -> None:
+        """
+        Handle an event on a connection socket.
+
+        You can simulate an EVENT_READ on in-memory data by setting mask==0 and data!=None.
+        """
+        assert self.sel
+        pid = self.conns[conn]
+        if mask & selectors.EVENT_READ:
+            data = conn.recv(4096)
+        # Messages are all one-byte codes for easy parsing.
+        # The protocol is "any number of keepalive "k"s followed by a quit "q", so we can
+        # safely ignore everything except the final byte of the message.
+        if data:
+            if data[-1:] == b"k":
+                # keepalive message; leave the connection alone.
+                return
+            elif data[-1:] == b"q":
+                # Graceful shutdown code.
+                self.graceful_shutdowns.append(pid)
+            else:
+                raise ValueError("invalid message from pid_client:", data)
+
+        # Error, EOF, or anything else.
```

Comment:
should this do something if this connection errored out, but the process stays alive indefinitely?

Reply:
Personally I think that is not a case that we reasonably need to be concerned with.  That would mean something like "the loopback interface broke".  I can add checks for it if you feel otherwise.